common/djangoapps/nodebb/signals/handlers.py

```diff
@@ -44,33 +50,55 @@ def handle_course_cert_awarded(sender, user, course_key, **kwargs):  # pylint: d
     send_user_course_completions_to_mailchimp.delay(data)
 
 
+@receiver(pre_save, sender=UserProfile)
+def user_profile_pre_save_callback(sender, **kwargs):
+    """
+    Capture old fields on the user_profile instance before save and cache them as a
+    private field on the current model for use in the post_save callback.
+    """
+    user_profile = kwargs['instance']
+    user_profile._updated_fields = get_changed_fields_dict(user_profile, sender)
+
+
 @receiver(post_save, sender=UserProfile)
-def sync_user_profile_info_with_nodebb(sender, instance, created, **kwargs):
-    city_of_residence = instance.city
-    country_of_residence = COUNTRIES.get(instance.country.code, '')
-    birth_year = instance.year_of_birth
-    language = instance.language
-
-    if not created and (country_of_residence or birth_year or language):
-        user = instance.user
-        data_to_sync = {
-            "city_of_residence": city_of_residence,
-            "country_of_residence": country_of_residence,
-            "birthday": "01/01/%s" % birth_year,
-            "language": language,
-        }
-        task_update_user_profile_on_nodebb.delay(
-            username=user.username, profile_data=data_to_sync)
+def sync_user_profile_info_with_nodebb(sender, instance, **kwargs):
+    updated_fields = getattr(instance, '_updated_fields', {})
+
+    if not ('city' in updated_fields or
+            'country' in updated_fields or
+            'year_of_birth' in updated_fields or
+            'language' in updated_fields):
+        return
+
+    user = instance.user
+    data_to_sync = {
+        "city_of_residence": instance.city,
+        "country_of_residence": COUNTRIES.get(instance.country.code, ''),
+        "birthday": "01/01/%s" % instance.year_of_birth,
+        "language": instance.language,
+    }
+    task_update_user_profile_on_nodebb.delay(
+        username=user.username, profile_data=data_to_sync)
 
 
 @receiver(post_save, sender=UserExtendedProfile)
 def sync_extended_profile_info_with_nodebb(sender, instance, **kwargs):
     request = get_current_request()
 
-    if not request:
+    user = instance.user
+
+    changed_fields = getattr(instance, '_changed_fields', {})
+
+    # return if fields to be updated on nodebb are not changed
+    if not request or not ('country_of_employment' in changed_fields or
```

Comment:
Nofal I see these checks in almost every signal, heres a way I think we can write them, which in my mind improves readability

I'll write down for this method, if it makes sense to you, you can update the rest as well


```
relevant_signal_fields = ['country_of_employment', 'city_of_employment']
relevant_signal_fields += instance.get_user_selected_interests(_type='field_name')
relevant_signal_fields += instance.get_user_selected_functions(_type='field_name')

if not request or not any([field in changed_field for field in relevant_signal_fields])

```

Reply:
it makes the check easy to read.

Reply:
Thanks @saad-waqar for awesome suggestion. I have made the changes and it looks much better.